### PR TITLE
Pass check_vjp, add VJP grad and Jac, use fun(y0, t, *args) syntax

### PR DESCRIPTION
**Pull request by cellweaver**
- Improved the VJP calculation so that it passes `jax.test_util.check_vjp()`. See `test_odeint_vjp()` for details.
- Added `build_odeint()` to define ODE integration functions with associated VJPs and `test_defvjp_all()` to test it.
- Added example functions of VJP gradient and Jacobian for `odeint()` using `vjp_odeint()`.
- Used `fun(y, t, *args)` syntax for the integrand system throughout. 
- Added tests that will be refactored into the proper place in `/tests/` in the next pull request once I figure out how the testing suite works.
- `decay()` hangs `check_vjp` so I have not yet included it as a test. `decay()` exhibits some instability.

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
This looks great!

On a JAX internals note, I think we might benefit from making this a Primitive rather than a custom_transforms function, but that's just a detail I'd like to look into in the future. (Using custom_transforms is the publicly-documented way to set up a Primitive, but for a core function like odeint I think we might want to make it a Primitive directly.)

I can move the tests in a follow-up commit.

Thanks for this contribution!

**Comment by mattjj** [comment]
Sorry for the long delay on our part!
